content/blog/2020/07/2020-07-27-machine-learning-plugin-coding-phase2.adoc

```diff
@@ -0,0 +1,160 @@
+---
+layout: post
+title: "Machine Learning Plugin project - Coding Phase 2 blog post"
+tags:
+- machinelearning
+- datascience
+- plugins
+- communitybonding
+- gsoc
+- gsoc2020
+author: loghijiaha
+description: "Coding Phase 2 of Google Summer of Code  2020"
+opengraph:
+  image: /images/gsoc/opengraph.png
+---
+
+image::/images/gsoc/jenkins-gsoc-logo_small.png[align="center"]
+
+Welcome back forks!
+
+This blog post is briefing my coding phase 2 in Jenkins link:/projects/gsoc/2020/projects/machine-learning/[Machine Learning Plugin] for this GSoC 2020.
+
+After successfully passed the evaluation and demo in the phase 1, our team were ahead for facing the challenges in phase 2.
+
+=== Summary
+
+This phase of coding was well spent with fixing lots of bugs and documentation. As the main feature of connecting to a IPython Kernel is done in the phase 1, we were able to focus on fixing minor/major bugs and documenting for the users. According to the JENKINS-62927 issue, a Docker agent was built to facilitate users without concerning plugin dependencies in python. In the act of deprecation of Python 2, we were indeed to port our plugin to support Python 3. We have tested our plugin in conda, venv and windows environments.
+Machine learning plugin has successfully passed the end to end test. A feature for a code editor is needed for further discussion/analysis as we have done a simple editor that may be useful in other ways in the future. link:https://github.com/jenkinsci/machine-learning-plugin/pull/35[PR#35]
+
+==== Main features of Machine Learning plugin
+
+- Run Jupyter notebook, (Zeppelin) JSON and Python files
+- Run Python code directly
+- Convert Jupyter Notebooks to Python and JSON
+- Configure IPython kernel properties
+- Support to execute Notebooks/Python on Agent
+- Support for Windows and Linux
+
+====== Upcoming features
+
+- Extract graph/map/images from the code
+- Save artifacts according to the step name
+- Generate reports for corresponding build
+
+====== Future improvements
+
+- Usage of JupyterRestClient
+- Support for multiple language kernels
+
+==== Docker agent
+
+The following dockerfile can be used to build the docker container as an agent for machine learning plugin. This docker agent can be used to run  notebooks or python scripts.
+
+.Dockerfile
+
+```
+FROM jenkins/agent:latest
+
+MAINTAINER Loghi <loghijiaha@gmail.com>
+
+USER root
+
+RUN apt update && apt install --no-install-recommends python3 -y \
+    python3-pip \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY requirements.txt /requirements.txt
+
+RUN pip3 install --upgrade pip setuptools && \
+    pip3 install --no-cache-dir -r /requirements.txt && \
+    ln -sf /usr/bin/python3 /usr/bin/python && \
+    ln -sf /usr/bin/pip3 /usr/bin/pip
+
+USER jenkins
+```
+
+
+==== Ported to Python 3
```

Comment:
Is this considered a subsection of the "Docker agent" section or is it an independent section in the blog post?
```suggestion
=== Ported to Python 3
```